crates/legacy/src/service.rs

```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
+            "The last built block should be retrievable"
+        );
+
+        let last_built_block = last_built_block.unwrap();
+
+        assert_eq!(
+            last_built_block.builder_hash, builder_hash,
+            "The last built block id should match the block id"
+        );
+
+        assert_eq!(
+            last_built_block.block_size, block_size_2,
+            "The last built block size should match the response message"
+        );
+        assert_ne!(
+            last_built_block.block_size, block_size_1,
+            "The last built block size should not match the previous block size"
+        );
+
+        assert_eq!(
+            last_built_block.offered_fee, offered_fee_2,
+            "The last built block offered fee should match the response message"
+        );
+        assert_ne!(
+            last_built_block.offered_fee, offered_fee_1,
+            "The last built block offered fee should not match the previous block offered fee"
+        );
+    }
+
+    // GlobalState::remove_handles Tests
+
+    /// This test checks to ensure that remove_handles will only consider
+    /// views up to what is known to have been stored. As a result it will
+    /// indicate that is has only targeted to the highest view number that it
+    /// is aware of.
+    #[async_std::test]
+    async fn test_global_state_remove_handles_prune_up_to_latest() {
+        let (bootstrap_sender, _) = async_broadcast::broadcast(10);
+        let (tx_sender, _) = async_broadcast::broadcast(10);
+        let parent_commit = vid_commitment(&[0], 4);
+        let mut state = GlobalState::<TestTypes>::new(
+            bootstrap_sender,
+            tx_sender,
+            parent_commit,
+            ViewNumber::new(0),
+            ViewNumber::new(0),
+            10,
+        );
+
+        // We register a few builder states.
+        for i in 1..=10 {
+            state.register_builder_state(
+                BuilderStateId {
+                    parent_commitment: vid_commitment(&[i], 4),
+                    view: ViewNumber::new(i as u64),
+                },
+                async_broadcast::broadcast(10).0,
+            );
+        }
+
+        assert_eq!(
+            state.remove_handles(ViewNumber::new(100)),
+            ViewNumber::new(10),
+            "It should only be able to prune up to what has been stored"
+        );
+
+        assert_eq!(
+            state.spawned_builder_states.len(),
+            1,
+            "The spawned_builder_states should only have a single entry in it"
+        );
+
+        for i in 0..10 {
+            let builder_state_id = BuilderStateId {
+                parent_commitment: vid_commitment(&[i], 4),
+                view: ViewNumber::new(i as u64),
+            };
+            assert!(
+                !state.spawned_builder_states.contains_key(&builder_state_id),
+                "the spawned builder states should not contain the builder state id, {builder_state_id}"
+            );
+        }
```

Comment:
Just checking if this comment was missed. :)

Reply:
I did miss this one!  Hmm... The previous entry does ensure that there is only one entry stored within the `spawned_builder_states`, however it doesn't test **which** entry remains.  This for-loop is testing to ensure that all of the entries that *should* have been removed are no longer preset within the storage.

We *could* potentially replace this with a check for the `i = 10`th entry only to achieve the same thing.  But I don't think checking for the non-existence of entries is really "bad".  If anything it helps to codify the assumptions, and ensure that our underlying storage is working correctly instead of just saying that it's working.

Reply:
> We _could_ potentially replace this with a check for the `i = 10`th entry only to achieve the same thing.

Yeah exactly, IIUC I think this is done via the `assert` at line 2676, which is why I think the loop here is a bit redundant. Nothing incorrect though.

Reply:
Addressed in 81e5660b8d7cf854a774dbc3514af3be67bc4fbe.